lsp/nls/src/requests/completion.rs

```diff
@@ -1,16 +1,227 @@
+    let mut seen: Vec<CompletionItem> = Vec::new();
+    for item in items {
+        if !seen.iter().any(|seen_item| seen_item.label == item.label) {
+            seen.push(item.clone())
+        }
+    }
+    seen
+}
+
+/// Search the linearization to find the record information associated with a
+/// partiular ID, and in the scope of a given linearization item.
+fn collect_record_info(
+    linearization: &Completed,
+    item: &LinearizationItem<Types>,
+    name: usize,
+) -> Vec<(Vec<Ident>, Types)> {
+    linearization
+        .get_in_scope(item)
+        .iter()
+        .filter_map(|item| {
+            let (ty, _) = linearization.resolve_item_type_meta(item);
+            match (&item.kind, ty) {
+                // Get record fields from static type info
+                (TermKind::Declaration(..), Types(AbsType::Record(row))) if name == item.id => {
+                    Some((extract_ident(&row), item.ty.clone()))
+                }
+                (TermKind::Declaration(_, _, ValueState::Known(body_id)), _) if name == item.id => {
+                    match find_record_contract_fields(&linearization, *body_id) {
+                        // Get record fields from contract metadata
+                        Some(fields) => Some((fields, item.ty.clone())),
+                        // Get record fields from lexical scoping
+                        None => find_record_fields(&linearization, *body_id)
+                            .map(|idents| (idents, item.ty.clone())),
+                    }
+                }
+                _ => None,
+            }
+        })
+        .collect()
+}
+
+/// Generate possible completion Identifiers given a source text, and its
+/// linearization.
+fn get_completion_identifiers(
+    source: &str,
+    trigger: Option<&str>,
+    linearization: &Completed,
+    item: &LinearizationItem<Types>,
+) -> Result<Vec<CompletionItem>, ResponseError> {
+    let in_scope = match trigger {
+        // Record Completion
+        Some(server::DOT_COMPL_TRIGGER) => {
+            let name = get_identifier_before_dot(source).ok_or(ResponseError {
+                code: ErrorCode::InternalError as i32,
+                message: "Couldn't get identifier for record completion".to_owned(),
+                data: None,
+            })?;
+            let name = Ident::from(name);
+            if let Some(name) = item.env.get(&name).cloned() {
+                collect_record_info(linearization, item, name)
+            } else {
+                return Ok(Vec::new());
+            }
+        }
+        Some(..) | None => {
+            // This is also record completion, but it is in the form
+            // <record var>.<partially-typed-field>
+            // we also want to give completion based on <record var> in this case.
+            if let Some(name) = get_identifier_before_field(source) {
+                let name = Ident::from(name);
+                if let Some(name) = item.env.get(&name).cloned() {
```

Comment:
```suggestion
                if let Some(name) = item.env.get(&name).copied() {
```

It's a nitpick, as both function are doing exactly the same thing. Still, it makes clear the fact that this is a very cheap operation: I was initially wondering while you need to clone a name, as you will probably just is it as a key or an index.